### PR TITLE
[Cart] Refactor removing cart item from cart

### PR DESCRIPTION
**Pull request by GSadee**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 1]
```diff
@@ -7,10 +7,29 @@ Feature: Removing cart item from cart
     Background:
         Given the store operates on a single channel in "France"
         And the store has a product "T-shirt banana" priced at "€12.54"
-        And I add this product to the cart
 
     @ui
     Scenario: Removing cart item
-        Given I see the summary of my cart
-        When I remove product "T-shirt banana" from the cart
+        Given I add this product to the cart
```
Comment (michalmarcinkowski) [inline features/cart/shopping_cart/removing_cart_item.feature]: `Given I added ""T-shirt banana" product to the cart`

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 2]
```diff
@@ -7,10 +7,29 @@ Feature: Removing cart item from cart
...
+        When I see the summary of my cart
```
Comment (michalmarcinkowski) [inline features/cart/shopping_cart/removing_cart_item.feature]: When I **view**~~see~~

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 3]
```diff
@@ -7,10 +7,29 @@ Feature: Removing cart item from cart
...
+        And I remove product "T-shirt banana" from the cart
+        Then my cart should be empty
+        And my cart's total should be "€0.00"
+
+    @ui
+    Scenario: Removing cart item when the store has defined default shipping method
+        Given the store has "UPS" shipping method with "€20.00" fee
+        And I add this product to the cart
```
Comment (michalmarcinkowski) [inline features/cart/shopping_cart/removing_cart_item.feature]: I add**ed** "T-shirt banana" to the cart

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 4]
```diff
@@ -7,10 +7,29 @@ Feature: Removing cart item from cart
...
+        When I see the summary of my cart
+        And I remove product "T-shirt banana" from the cart
         Then my cart should be empty
+        And my cart's total should be "€0.00"
+
+    @ui
+    Scenario: Checking cart's total after removing one item
+        Given I add this product to the cart
```
Comment (NeverResponse) [inline features/cart/shopping_cart/removing_cart_item.feature]: Same step in every scenario, wouldn't it be better to move it to the background?

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 5]
```diff
@@ -7,10 +7,29 @@ Feature: Removing cart item from cart
...
+        And the store has a product "T-shirt strawberry" priced at "€17.22"
+        And I add this product to the cart
+        When I see the summary of my cart
+        And I remove product "T-shirt banana" from the cart
+        And my cart's total should be "€17.22"
```
Comment (michalmarcinkowski) [inline features/cart/shopping_cart/removing_cart_item.feature]: This scenario should be the following:
```
Given the store also has a product "T-shirt strawberry" priced at "€17.22"
And I added this product to the cart
And I also added "T-shirt banana" to the cart
When I remove product "T-shirt strawberry" from my cart
Then my cart's total should be "€12.54"
```

File: src/Sylius/Behat/Context/Ui/Shop/CartContext.php [hunk 6]
```diff
@@ -449,4 +449,15 @@ public function thisItemPriceShouldBe($productName, $productPrice)
...
     }
+
+    /**
+     * @Then my cart's total should be :total
+     */
+    public function myCartSTotalShouldBe($total)
+    {
+        Assert::true(
+            $this->summaryPage->hasCartTotal($total),
+            sprintf('Cart total should have %s total, but it does not have.', $total)
```
Comment (NeverResponse) [inline src/Sylius/Behat/Context/Ui/Shop/CartContext.php]: Better yet, grab the total and compare, so we can have a more verbose message.

File: src/Sylius/Component/Core/Model/Order.php [hunk 7]
```diff
@@ -346,6 +346,16 @@ public function removeShipment(ShipmentInterface $shipment)
     /**
      * {@inheritdoc}
      */
+    public function removeShipments()
+    {
+        foreach ($this->getShipments() as $shipment) {
+            $this->removeShipment($shipment);
+        }
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/Order.php]: The following implementation will be fine: `$this->shipments->clear();`

File: src/Sylius/Component/Core/OrderProcessing/OrderShipmentProcessor.php [hunk 8]
```diff
@@ -69,6 +69,12 @@ public function process(OrderInterface $order)
      */
     private function getOrderShipment(OrderInterface $order)
     {
+        if ($order->isEmpty()) {
+            $order->removeShipments();
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/OrderProcessing/OrderShipmentProcessor.php]: We shouldn't do it in `getOrderShipment` method but rather in the beginning of a `process`method.

**Comment by pjedrzejewski** [comment]
@GSadee Requires rebasing.

**Comment by michalmarcinkowski** [comment]
Good job! 👍

**Pull request merged**